tests/test__collection_api.py

```diff
@@ -2597,6 +2597,36 @@ def test__aggregate_arithmetic(self):
             [{'sum': 20.5, 'prod': 30, 'trunc': 1}],
             [{k: v for k, v in doc.items() if k != '_id'} for doc in actual])
 
+    def test__aggregate_string_operations(self):
+        self.db.collection.insert_one({
+            'a': 'Hello',
+            'b': 'World',
+        })
+        actual = self.db.collection.aggregate([{'$project': {
+            'concat': {'$concat': ['$a', ' Dear ', '$b']},
+            'sub1': {'$substr': ['$a', 0, 4]},
+            'sub2': {'$substr': ['$a', -1, 3]},
+            'lower': {'$toLower': '$a'},
+            'upper': {'$toUpper': '$a'},
+        }}])
+        self.assertEqual(
+            [{'concat': 'Hello Dear World', 'sub1': 'Hell', 'sub2': '', 'lower': 'hello', 'upper': 'HELLO'}],
+            [{k: v for k, v in doc.items() if k != '_id'} for doc in actual])
+
+    def test__aggregate_misc_operations(self):
+        _id = ObjectId()
```

Comment:
Please use an ObjectId for which you choose the hash:
```py
_id = ObjectId('5abcfad1fbc93d00080cfe66')

...

assertEqual([{'toString1: '5abcfad1fbc93d00080cfe66', 'toString2': '3'}], ...)
```

Reply:
The rationale is that I prefer to keep the logic in tests to their minimum.

Reply:
good idea. np